pkg/provisioners/kafka/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package dispatcher
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+	"github.com/bsm/sarama-cluster"
+	"github.com/google/go-cmp/cmp"
+	"go.uber.org/zap"
+
+	eventingduck "github.com/knative/eventing/pkg/apis/duck/v1alpha1"
+	"github.com/knative/eventing/pkg/buses"
+	"github.com/knative/eventing/pkg/provisioners/kafka/controller"
+	"github.com/knative/eventing/pkg/sidecar/multichannelfanout"
+)
+
+type KafkaDispatcher struct {
+	config     atomic.Value
+	updateLock sync.Mutex
+
+	receiver   *buses.MessageReceiver
+	dispatcher *buses.MessageDispatcher
+
+	kafkaAsyncProducer sarama.AsyncProducer
+	kafkaConsumers     map[buses.ChannelReference]map[subscription]KafkaConsumer
+	kafkaCluster       KafkaCluster
+
+	logger *zap.Logger
+}
+
+type KafkaConsumer interface {
+	Messages() <-chan *sarama.ConsumerMessage
+	MarkOffset(msg *sarama.ConsumerMessage, metadata string)
+	Close() (err error)
+}
+
+type KafkaCluster interface {
+	NewConsumer(groupID string, topics []string) (KafkaConsumer, error)
+}
+
+type saramaCluster struct {
+	kafkaBrokers []string
+}
+
+func (c *saramaCluster) NewConsumer(groupID string, topics []string) (KafkaConsumer, error) {
+	consumerConfig := cluster.NewConfig()
+	consumerConfig.Version = sarama.V1_1_0_0
+	return cluster.NewConsumer(c.kafkaBrokers, groupID, topics, consumerConfig)
+}
+
+type subscription struct {
+	Namespace     string
+	Name          string
+	SubscriberURI string
+	ReplyURI      string
+}
+
+// ConfigDiffs diffs the new config with the existing config. If there are no differences, then the
+// empty string is returned. If there are differences, then a non-empty string is returned
+// describing the differences.
+func (d *KafkaDispatcher) ConfigDiff(updated *multichannelfanout.Config) string {
+	return cmp.Diff(d.getConfig(), updated)
+}
+
+func (d *KafkaDispatcher) UpdateConfig(config *multichannelfanout.Config) error {
+	if config == nil {
+		return errors.New("nil config")
+	}
+
+	d.updateLock.Lock()
+	defer d.updateLock.Unlock()
+
+	if diff := d.ConfigDiff(config); diff != "" {
```

Comment:
Hmm, if you have multiple replicas, won't they each attempt to consume from the same consumer group? Will all but one block (I'm not familiar enough with Kafka and this library to be sure, it's possible that this is simply a competing consumer model)?

Obviously, with only 1 partition, the benefit of replicas will be limited, but I'm guessing that we'll support wider fanout soon.

Reply:
The [sarama-cluster](https://godoc.org/github.com/bsm/sarama-cluster) library provides a balanced consumer group implementation on top of [sarama](https://godoc.org/github.com/Shopify/sarama). Even with multiple replicas, since consumer instances are part of the same consumer group, they coordinate so that at most one consumer will have access to a partition in a topic subscribed by the group.

> Obviously, with only 1 partition, the benefit of replicas will be limited, but I'm guessing that we'll support wider fanout soon.

We already support `NumPartitions` as a channel argument.